src/main/java/org/manifold/compiler/front/Main.java

```diff
@@ -376,45 +356,88 @@ public Expression visitTupleValue(TupleValueContext context) {
         identifier = nextAnonymousID.toString();
         nextAnonymousID += 1;
       }
-      valueExprs.put(identifier, value);
+      ExpressionEdge eValue = new ExpressionEdge(vxValue, null);
+      valueEdges.put(identifier, eValue);
+      exprGraph.addEdge(eValue);
     }
-    return new TupleValueExpression(valueExprs);
+    TupleValueVertex vTuple = new TupleValueVertex(exprGraph, valueEdges);
+    exprGraph.addVertex(vTuple);
+    return vTuple;
   }
 
   @Override
-  public Expression visitFunctionTypeValue(FunctionTypeValueContext context) {
-    Expression inputExpr = context.tupleTypeValue(0).accept(this);
-    Expression outputExpr = context.tupleTypeValue(1).accept(this);
-    return new FunctionTypeValueExpression(inputExpr, outputExpr);
+  public ExpressionVertex visitFunctionTypeValue(
+      FunctionTypeValueContext context) {
+    // get the vertex corresponding to the input type
+    ExpressionVertex vIn = context.tupleTypeValue(0).accept(this);
+    ExpressionEdge eIn = new ExpressionEdge(vIn, null);
+    // then get the output type vertex
+    ExpressionVertex vOut = context.tupleTypeValue(1).accept(this);
+    ExpressionEdge eOut = new ExpressionEdge(vOut, null);
+    
+    FunctionTypeValueVertex vFunctionType = new FunctionTypeValueVertex(
+        exprGraph, eIn, eOut);
+    exprGraph.addVertex(vFunctionType);
+    exprGraph.addEdge(eIn);
+    exprGraph.addEdge(eOut);
+    return vFunctionType;
   }
 
   @Override
-  public Expression visitNamespacedIdentifier(
+  public ExpressionVertex visitNamespacedIdentifier(
       NamespacedIdentifierContext context) {
-
+    // keeping in mind that we may have constructed this variable already...
     List<TerminalNode> identifierNodes = context.IDENTIFIER();
     List<String> identifierStrings = new LinkedList<>();
     for (TerminalNode node : identifierNodes) {
       identifierStrings.add(node.getText());
     }
 
-    VariableIdentifier variable = new VariableIdentifier(identifierStrings);
-
-    return new VariableReferenceExpression(variable);
+    VariableIdentifier id = new VariableIdentifier(identifierStrings);
+    if (ReservedIdentifiers.getInstance()
+        .isReservedIdentifier(id)) {
+      // construct a constant value vertex with the identifier's value
+      ConstantValueVertex vReserved = new ConstantValueVertex(exprGraph,
+          ReservedIdentifiers.getInstance().getValue(id));
+      exprGraph.addVertex(vReserved);
+      return vReserved;
+    } else {
+      // this is a variable
+      // TODO scope
+      try {
```

Comment:
Yikes, exception handling for control flow is generally a bad smell. Perhaps we need something like ```exprGraph.containsVariable(VariableIdentifier)```? Then this can be just something like:

```
try {
    if (!exprGraph.containsVariable(id) {
        exprGraph.addVertex(id);
    }
    return exprGraph.getVariableVertex(id);
} catch (MultipleDefinitionException | VariableNotDefinedException e) {
    // should never happen
    throw Throwables.propagage(e);
}
```